aws-sdk-resources/lib/aws-sdk-resources/services/s3/file_downloader.rb

```diff
@@ -135,31 +133,44 @@ def map_files(slice, dir, mode)
         end
       end
 
-      def get_range(chunks)
+      def clean_up_parts(parts)
+        parts.each do |filename|
+          File.unlink(filename) if File.exists?(filename)
+        end
+      end
+
+      def multithreaded_get_by_ranges(chunks)
         thread_batches(chunks, 'range')
       end
 
-      def get_part(parts)
+      def multithreaded_get_by_parts(parts)
         thread_batches(parts, 'part_number')
       end
 
       def thread_batches(chunks, param)
         batches = file_batches(chunks, param)
+        parts = batches.inject([]) {|a, batch| a.push(*batch.values); a}
         batches.each do |batch|
           threads = []
           batch.each do |chunk, file|
-            threads << Thread.new do
-              resp = @client.get_object(
-                :bucket => @bucket,
-                :key => @key,
-                param.to_sym => chunk,
-                :response_target => file
-              )
+            threads << thread = Thread.new do
+              begin
+                resp = @client.get_object(
+                  :bucket => @bucket,
+                  :key => @key,
+                  param.to_sym => chunk,
+                  :response_target => file
+                )
+              rescue => error
+                # clear file parts once failed
+                clean_up_parts(parts)
+                raise error
+              end
+              thread.abort_on_exception = true
```

Comment:
Good catch on setting `Thread#abort_on_exception`, I think this `ObjectMultipartCopier` and `MultipartFileUploader` should also be updated to set `Thread#abort_on_exception` (otherwise the thread won't propagate the error, and nothing will be cleaned up).

Btw, you probably need to set `Thread#abort_on_exception` *before* the exception is raised (at the top of the thread definition), otherwise that part of code will never get executed in case an exception is raised. While here, instead of using a `thread` local variable, you could also use `Thread.current` to access the current thread:

```rb
threads << Thread.new do
  Thread.current.abort_on_exception = true

  begin
    # execute
  rescue
    # clean-up
  end
end
```